### DISCUSSION TITLE
Switch from Lua to more-modern language (MoonScript / TypeScript)

### DISCUSSION THREAD

**Discussion by Cheatoid** [vurvdev/VExtensions#24] [Ideas]
This is an interesting idea I'd like to explore (in future).
Garry's Mod addons use [Lua](https://www.lua.org) language for scripting (technically *LuaJIT*, but that's irrelevant). **But why stop there?**
There's a branch of compilers that are able to read the source code (written in different language) and generate the code for the target language (in this case, Lua). Technically called ["transpilers", more on Wiki](https://en.wikipedia.org/wiki/Source-to-source_compiler).

What is even more exciting, it is that the community is aware of this and they already take advantage of it! 😄
For example, check out [ULX 4](https://github.com/TeamUlysses/ulx4), it uses [MoonScript](https://moonscript.org).
And there's more, check out [Among Us](https://github.com/NotMyWing/GarrysModAmongUs) (gamemode) which also uses [MoonScript](https://moonscript.org).
I haven't come across an addon which uses [TypeScript](https://github.com/TypeScriptToLua/TypeScriptToLua) yet.

While this would require a complete rewrite (more or less), there are a number of advantages to do this.

**Comment by vurvdev** [comment]
Sounds interesting, but I really don't think this would apply for something that is just an extension to Wiremod, maybe if this was a separate, bigger project like a fork of Wiremod I would be more open to this.

And it's a hard no for Moonscript, it just looks really weird to me and looks like a lot of bloat in comparison to a lua implementation when I'm looking at https://github.com/TeamUlysses/ulx4/blob/master/moon/lib/tableshape.moon. It also seems to have no practical use besides for more "easy" object/class management, even if it is pretty easy to do with lua already. TypescriptToLua looks neat though.

Also found more interesting ones here https://github.com/hengestone/lua-languages, includes C# to lua, Go to lua and a ton of obscure ones

Edit: The only one I'd consider for now is wu, I'm kind of hooked on low-level like languages and I like how it behaves like Rust https://github.com/wu-lang/wu

**Comment by Cheatoid** [comment]
Exploring *CSharp.lua* and *wu*...
They look interesting, personally I would prefer C#. I have managed to get it up and running in GMod, still needs some work in order to call GMod Lua (API) functions, looking into that now...
Me being crazy enough, I think I will make this compatible with Starfall too. 💯

There is only 1 problem if we were to switch to a different language, the E2 DSL, lua files where `e2function`s are defined at, it is not regular Lua, but that can be workarounded by registering E2 functions in old-fashion way (but we would make a nicer wrapper for it beforehand; the [`registerFunction`](https://github.com/wiremod/wire/search?q=registerFunction)) -or- the lazy workaround: just ignore switching language for such files and keep them as they are.

**Comment by vurvdev** [comment]
https://lib.haxe.org/p/gmodhaxe/

**Comment by vurvdev** [comment]
https://github.com/Vurv78/VExtensions/tree/typescript
